pkg/ottl/ottlfuncs/func_index.go

```diff
@@ -0,0 +1,153 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+var valueComparator = ottl.NewValueComparator()
+
+type IndexArguments[K any] struct {
+	Source ottl.Getter[K]
```

Comment:
```suggestion
	Target ottl.Getter[K]
```

plus docs, so it follows the most common pattern.